velox/functions/lib/SIMDComparisonUtil.h

```diff
@@ -233,70 +239,189 @@ void applyAutoSimdComparison(
-        [&](const A* __restrict rawA, const B* __restrict rawB, int i) {
+        [&](const A* __restrict rawA,
+            const B* __restrict rawB,
+            const vector_size_t* __restrict indexA,
+            const vector_size_t* __restrict indexB,
+            int i) {
           if constexpr (sizeof...(cmpArgs) > 0) {
             return Compare::apply(rawA[0], rawB[i], cmpArgs...);
           } else {
             return cmp(rawA[0], rawB[i]);
           }
         },
-        result);
+        result,
+        rows,
+        rawA,
+        rawB,
+        nullptr,
+        nullptr);
   } else if (args[0]->isFlatEncoding() && args[1]->isConstantEncoding()) {
     const A* __restrict rawA =
         args[0]->asUnchecked<FlatVector<A>>()->template rawValues<A>();
     const B constB = args[1]->asUnchecked<ConstantVector<B>>()->valueAt(0);
     const B* __restrict rawB = &constB;
     detail::applyAutoSimdComparisonInternal(
-        rows,
-        rawA,
-        rawB,
-        [&](const A* __restrict rawA, const B* __restrict rawB, int i) {
+        [&](const A* __restrict rawA,
+            const B* __restrict rawB,
+            const vector_size_t* __restrict indexA,
+            const vector_size_t* __restrict indexB,
+            int i) {
           if constexpr (sizeof...(cmpArgs) > 0) {
             return Compare::apply(rawA[i], rawB[0], cmpArgs...);
           } else {
             return cmp(rawA[i], rawB[0]);
           }
         },
-        result);
+        result,
+        rows,
+        rawA,
+        rawB,
+        nullptr,
+        nullptr);
   } else if (args[0]->isConstantEncoding() && args[1]->isConstantEncoding()) {
     const A constA = args[0]->asUnchecked<ConstantVector<A>>()->valueAt(0);
     const A* __restrict rawA = &constA;
     const B constB = args[1]->asUnchecked<ConstantVector<B>>()->valueAt(0);
     const B* __restrict rawB = &constB;
     detail::applyAutoSimdComparisonInternal(
+        [&](const A* __restrict rawA,
+            const B* __restrict rawB,
+            const vector_size_t* __restrict indexA,
+            const vector_size_t* __restrict indexB,
+            int i) {
+          if constexpr (sizeof...(cmpArgs) > 0) {
+            return Compare::apply(rawA[0], rawB[0], cmpArgs...);
+          } else {
+            return cmp(rawA[0], rawB[0]);
+          }
+        },
+        result,
         rows,
         rawA,
         rawB,
-        [&](const A* __restrict rawA, const B* __restrict rawB, int i) {
+        nullptr,
+        nullptr);
+  } else if (
+      args[0]->isFlatEncoding() &&
+      args[1]->encoding() == VectorEncoding::Simple::DICTIONARY &&
+      args[1]
+          ->asUnchecked<DictionaryVector<B>>()
+          ->valueVector()
+          ->isFlatEncoding()) {
+    auto dictB = args[1]->asUnchecked<DictionaryVector<B>>();
+    const A* __restrict rawA =
+        args[0]->asUnchecked<FlatVector<A>>()->rawValues();
+    const B* __restrict rawB = dictB->valueVector()
```

Comment:
Just use `LocalDecodedVector` for `args[1]`, then you can remove the encoding restriction on `valueVector`

Reply:
Thanks, I changed to use LocalDecodedVector instead. Since I only observe the case with Dict(Flat), I would like to apply this only for the Dict(Flat) case.

Reply:
I think we should apply it for all cases when one argument is dictionary encoded.  Can you think about one counter example that we should not do that?

Reply:
I see your point, no counter example form my side. Updated to cover all cases  when one argument is dictionary and one is flat.
Thanks.